### PR TITLE
Add zfs pool stats collection.

### PR DESCRIPTION
**Pull request by allenpetersen**
Add optional support for gathering zfs pool iostats.

### CODE REVIEW

**Comment by allenpetersen** [comment]
I noticed there was an issue with the original zfs test where it really wasn't testing a single dataset. The second commit addresses this.

File: plugins/zfs/zfs.go [hunk 1]
```diff
@@ -57,7 +103,10 @@ func (z *Zfs) Gather(acc plugins.Accumulator) error {
 		kstatPath = "/proc/spl/kstat/zfs"
 	}
 
-	tags := getTags(kstatPath)
+	tags, err := getPoolStats(kstatPath, z.PoolMetrics, acc)
```
Comment (sparrc) [inline plugins/zfs/zfs.go]: I don't understand, why would `getPoolStats` return tags for the metrics?

File: plugins/zfs/zfs.go [hunk 2]
```diff
@@ -32,7 +37,7 @@ func (z *Zfs) Description() string {
 	return "Read metrics of ZFS from arcstats, zfetchstats and vdev_cache_stats"
 }
 
-func getTags(kstatPath string) map[string]string {
```
Comment (sparrc) [inline plugins/zfs/zfs.go]: Please keep `getTags` and make `getPoolStats` a separate function

File: plugins/zfs/zfs.go [hunk 3]
```diff
@@ -42,8 +47,49 @@ func getTags(kstatPath string) map[string]string {
...
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return map[string]string{"pools": pools}, nil
+}
+
+func readPoolStats(poolIoPath string, poolName string, acc plugins.Accumulator) error {
```
Comment (sparrc) [inline plugins/zfs/zfs.go]: change this function name, `readPoolStats` suggests that you are not adding metrics to the accumulator here. Maybe something like `gatherPoolStats`

**Comment by sparrc** [comment]
thanks @allenpetersen! Just a few comments to address and I can get this merged

File: plugins/zfs/zfs.go (hunk 2 above)
Comment (allenpetersen) [inline plugins/zfs/zfs.go]: I know it isn't ideal, I wanted to avoid doing the directory scan a second time for each Gather call.

Comment (sparrc) [inline plugins/zfs/zfs.go]: OK, I see what you mean, one thing you could do is make `tags` a private variable on the `Zfs` struct. Then you can set `z.tags["pool"] = ...` within this function, and don't return a map in that case.

You can either pass in the `Zfs` struct or make this function have a receiver `(z *Zfs)`

Comment (allenpetersen) [inline plugins/zfs/zfs.go]: I was thinking about putting the directory scan into a getPools() that returns a slice of structs containing the pool name and io path. Then the getPoolsTag() and gatherPoolStats() will both take this this as the input. Do you like that better than adding a tags property to Zfs?

Comment (sparrc) [inline plugins/zfs/zfs.go]: that works as well, :+1:

**Comment by allenpetersen** [comment]
@sparrc I think this cleans things up. I didn't love my first pass but I was trying to minimize the amount of change to the existing code.

**Comment by sparrc** [comment]
thanks again @allenpetersen!

**Pull request closed without merging**